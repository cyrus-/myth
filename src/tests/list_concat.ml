let s = "
type nat =
  | O
  | S of nat

type list =
  | Nil
  | Cons of nat * list

type llist =
  | LNil
  | LCons of list * llist

let rec append (l1:list) (l2:list) : list =
  match l1 with
  | Nil -> l2
  | Cons (x, l1p) -> Cons (x, append l1p l2)
;;

let list_concat : llist -> list |>
  { LNil => []
  | LCons ([], LNil) => []
  | LCons ([0], LNil) => [0]
  | LCons ([0], LCons([0], LNil)) => [0;0]
  | LCons ([1], LNil) => [1]
  | LCons ([1], LCons([1], LNil)) => [1;1]
  } = ?"
