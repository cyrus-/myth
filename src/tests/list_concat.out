(* Generated Thu Nov 13 00:35:23 2014 *)
let list_concat : llist -> list =
  let rec f1 (l1:llist) : list =
    match l1 with
      | LNil -> Nil
      | LCons (l2, l3) -> append l2 (f1 l3)
  in
    f1
;;
